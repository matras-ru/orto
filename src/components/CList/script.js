import { selfInstall } from '@/utils';
import { mergeData } from 'vue-functional-data-merge';

import ThemeClass from '@/themes/default/CList';

const { baseClass, defaultClass, horizontalClass, startClass, endClass, betweenClass } = ThemeClass; // TODO: flex-class вынести

const props = {
    tag: {
        type: String,
        default: 'ul'
    },

    horizontal: {
        type: Boolean,
        default: false
    },

    justify: {
        type: String,
        default: null
    }
};

const currentClass = props => {
    let classes = [...baseClass];

    // horizontal/vertical
    classes.push(props.horizontal ? [...horizontalClass] : [...defaultClass]);

    // horizontal align
    switch (props.justify) {
        case 'end':
            classes.push(...endClass);
            break;
        case 'between':
            classes.push(...betweenClass);
            break;
        case 'start':
            classes.push(...startClass);
            break;
        default:
            break;
    }

    return classes;
};

export default {
    name: 'List',

    functional: true,

    install(Vue, theme) {
        selfInstall(Vue, theme, this);
    },

    props,

    render(h, { props, data, children }) {
        const componentData = {
            class: currentClass(props)
        };

        return h(props.tag, mergeData(data, componentData), children);
    }
};
