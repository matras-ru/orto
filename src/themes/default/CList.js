// TODO: вынести глобальные унивирсальные классы типо 'betweenClass' ???

const List = {
    baseClass: ['flex', 'flex-wrap'],
    defaultClass: ['flex-col'],
    horizontalClass: ['flex-row'],
    betweenClass: ['justify-between'],
    startClass: ['justify-start'],
    endClass: ['justify-end']
};

export default List;
