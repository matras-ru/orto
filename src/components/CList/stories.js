import { storiesOf } from '@storybook/vue';

import CList from './';
import CListItem from '../CListItem';

storiesOf('CList - CListItem', module)
    .add('vertical - default', () => ({
        components: { CList, CListItem },
        template: `
        <CList>
          <CListItem class="mb-6">Lorem, ipsum.</CListItem>
          <CListItem>Lorem ipsum dolor sit.</CListItem>
        <CList>
        `
    }))
    .add('horizontal', () => ({
        components: { CList, CListItem },
        template: `
        <CList horizontal>
          <CListItem>Lorem, ipsum.</CListItem>
          <CListItem class="ml-6">Lorem ipsum dolor sit.</CListItem>
        <CList>
        `
    }));
