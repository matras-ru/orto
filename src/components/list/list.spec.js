/* eslint-env jest */
import { mount } from '@vue/test-utils';
import CList from './list';

const baaseClass = 'flex';
const defaultCLass = 'flex-col';
const horizontalClass = 'flex-row';
const betweenClass = 'justify-between';

describe('CList', () => {
    it('default empty CList is functional component and rendered. Should have tag ul by default', () => {
        const wrapper = mount(CList);
        expect(wrapper.isFunctionalComponent).toBe(true);
        expect(wrapper.element.tagName).toEqual('UL');
        expect(wrapper.classes().sort()).toEqual(`${baaseClass} ${defaultCLass}`.split(' ').sort());

        expect(wrapper.html()).toMatchSnapshot();

        wrapper.destroy();
    });

    it('html tag equal div when prop tag = div', () => {
        const wrapper = mount(CList, {
            context: {
                props: {
                    tag: 'div'
                }
            }
        });

        expect(wrapper.element.tagName).toEqual('DIV');
        expect(wrapper.html()).toMatchSnapshot();

        wrapper.destroy();
    });

    it('horizontal mode', () => {
        const wrapper = mount(CList, {
            context: {
                props: {
                    direction: 'horizontal'
                }
            }
        });

        expect(wrapper.classes().sort()).toEqual(
            `${baaseClass} ${horizontalClass}`.split(' ').sort()
        );
        expect(wrapper.html()).toMatchSnapshot();

        wrapper.destroy();
    });

    it('between horizontal align', () => {
        const wrapper = mount(CList, {
            context: {
                props: {
                    justify: 'between'
                }
            }
        });

        expect(wrapper.classes().sort()).toEqual(
            `${baaseClass} ${defaultCLass} ${betweenClass}`.split(' ').sort()
        );
    });

    it('custom attributes', () => {
        const wrapper = mount(CList, {
            attrs: {
                role: 'tabs'
            }
        });

        expect(wrapper.attributes('role')).toBe('tabs');
    });
});
