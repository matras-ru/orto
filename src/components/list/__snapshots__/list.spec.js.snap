// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CList default empty CList is functional component and rendered. Should have tag ul by default 1`] = `<ul class="flex flex-col"></ul>`;

exports[`CList horizontal mode 1`] = `<ul class="flex flex-row"></ul>`;

exports[`CList html tag equal div when prop tag = div 1`] = `<div class="flex flex-col"></div>`;
